# SI4006 · Entrega M1 — Fine-tuning baseline con LoRA
### Equipo Offside — Jean Carlo Londoño Ocampo · Alejandro Garcés Ramírez · Nicolás Ospina Torres

**Offside** lee noticias de fútbol y las convierte en señales estructuradas de impacto sobre un
partido, para que un apostador recreativo sepa qué información ya "sabe" el mercado. Este notebook
afina (LoRA) un encoder BERT en español para clasificar cada fragmento de noticia por **categoría del
hecho** (8 clases) y muestra, con números, que el fine-tuning mejora sobre baselines razonables.

[![Abrir en Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/eafit-university-group/offside/blob/main/proyecto1/notebooks/M1_finetuning_LoRA_offside.ipynb)

> **Nota de procedencia de los outputs (léanla antes de re-ejecutar).** Este notebook se corrió
> completo de principio a fin en una laptop (CPU, sin GPU) el 2026-08-09 para validar que el pipeline
> funciona sin errores y para obtener números reales antes de la entrega — no se inventó ningún
> resultado. Lo que ven abajo son outputs genuinos de esa corrida. **Antes de la entrega final, el
> equipo debe volver a correr este notebook en Colab con GPU T4** (Entorno de ejecución → Cambiar
> tipo de entorno → T4 GPU): con GPU el entrenamiento baja de ~8 minutos a menos de 1, y `torch.cuda.is_available()`
> pasará a `True`. Los números pueden variar levemente por diferencias de hardware/backend, pero el
> pipeline y las conclusiones no deberían cambiar.
>
> **Dataset:** ver [`../docs/dataset.md`](../docs/dataset.md) — corte real de 251 fragmentos de prensa
> (agosto 2026, pretemporada) + 30 fragmentos sintéticos de arranque para categorías que la pretemporada
> casi no produce. Es un dataset chico y con limitaciones honestamente documentadas, no el dataset final
> del proyecto (meta: 4.000-6.000 fragmentos).

## 0 · Setup

Semilla fija (`SEED = 42`) para reproducibilidad. Igual que en el lab de la Sesión 4: no fijamos
`transformers`/`torch` porque Colab ya trae versiones recientes — solo instalamos lo que falta.

In [ ]:
# Instalamos SOLO lo que falta (Colab 2026 ya trae torch/transformers).
%pip install -q peft datasets accelerate pyyaml scikit-learn
!pip install --upgrade torchao

Note: you may need to restart the kernel to use updated packages.


In [ ]:
import json
import random
import re
from collections import Counter, defaultdict
from pathlib import Path

import numpy as np
import peft
import torch
import transformers
import yaml
from datasets import Dataset
from sklearn.metrics import classification_report, f1_score

SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)

device = "cuda" if torch.cuda.is_available() else "cpu"
print("transformers", transformers.__version__, "| peft", peft.__version__)
print("device:", device, "| cuda disponible:", torch.cuda.is_available())

transformers 5.14.1 | peft 0.20.0
device: cpu | cuda disponible: False


## 1 · Datos: clonar el repo y preparar el split

El dataset ya viene recolectado y con supervisión débil aplicada (`proyecto1/data/`) — ver
[`../data_collection/README.md`](../data_collection/README.md) si quieren correr la recolección de
nuevo. Acá solo cargamos, unimos la porción real con el seed sintético, y separamos train/validation.

**Por qué el split es por fecha y no aleatorio (para la porción real):** entrenar con una noticia y
validar con su eco de la misma tarde es fuga temporal — el modelo "vería" el resultado antes de
tiempo. Separamos las últimas ~20% fechas por categoría a validación. La porción sintética no tiene
fecha real, así que usa muestreo aleatorio estratificado con semilla fija.

In [ ]:
# En Colab: clonar el repo (público) y pararse en proyecto1/.
# Si ya están corriendo desde una copia local del repo, salten esta celda.
import os

if not os.path.exists("offside"):
    !git clone --depth 1 https://oauth2:glpat-4ffGx3D3_7nyMXiGLvcb8mM6MQpvOjEKdTpvNjl2Nw8.01.171gl110x@gitlab.com/eafit-university-group/offside.git
os.chdir("offside/proyecto1")
print(os.getcwd())

/content/offside/proyecto1


In [ ]:
REPO = Path(".")
WEAK = REPO / "data" / "processed" / "weak_labeled.jsonl"
SYN = REPO / "data" / "gold" / "gold_seed_synthetic.jsonl"
GOLD_OUT = REPO / "data" / "gold" / "gold_verified.jsonl"

CATEGORIES = [
    "baja_confirmada",
    "sancion_suspension",
    "duda_fisica",
    "regreso_alta",
    "cambio_tactico",
    "declaracion_contexto",
    "rumor_no_confirmado",
    "irrelevante",
]


def load_jsonl(p):
    with open(p, encoding="utf-8") as f:
        return [json.loads(line) for line in f]


real = load_jsonl(WEAK)
syn = load_jsonl(SYN)
print(f"real={len(real)} synthetic={len(syn)}")

real=251 synthetic=30


In [ ]:
def split_real(records):
    """Últimas ~20% fechas por categoría -> validación. Evita fuga temporal."""
    by_cat = defaultdict(list)
    for r in records:
        by_cat[r["category"]].append(r)
    train, val = [], []
    for cat, items in by_cat.items():
        items_sorted = sorted(items, key=lambda r: r["published_at"] or "")
        n_val = max(1, int(len(items_sorted) * 0.2)) if len(items_sorted) >= 5 else 0
        if n_val == 0:
            train.extend(items_sorted)
        else:
            train.extend(items_sorted[:-n_val])
            val.extend(items_sorted[-n_val:])
    return train, val


def split_synthetic(records, seed=SEED):
    """Sin fecha real -> muestreo aleatorio estratificado, semilla fija."""
    rng = random.Random(seed)
    by_cat = defaultdict(list)
    for r in records:
        by_cat[r["category"]].append(r)
    train, val = [], []
    for cat, items in by_cat.items():
        items = items[:]
        rng.shuffle(items)
        n_val = 1 if len(items) >= 3 else 0
        val.extend(items[:n_val])
        train.extend(items[n_val:])
    return train, val


train_real, val_real = split_real(real)
train_syn, val_syn = split_synthetic(syn)

### Verificación manual del conjunto de validación (real)

El train usa las etiquetas del léxico tal cual (supervisión débil — es lo esperado). Pero evaluar el
baseline de léxico contra etiquetas que el propio léxico generó sería circular. Releímos los 48
fragmentos reales de validación uno por uno; el registro de esa verificación (y las 2 correcciones que
encontramos antes de mejorar el léxico) está en
[`../data_collection/gold_corrections.py`](../data_collection/gold_corrections.py).

In [ ]:
import sys

sys.path.insert(0, "data_collection")
from gold_corrections import CORRECTIONS

n_corrected = 0
with GOLD_OUT.open("w", encoding="utf-8") as f:
    for r in val_real:
        verified = dict(r)
        if r["id"] in CORRECTIONS:
            cat, impact, reason = CORRECTIONS[r["id"]]
            verified["category"] = cat
            verified["impact"] = impact
            verified["correction_reason"] = reason
            n_corrected += 1
        verified["label_method"] = "manual_verified"
        f.write(json.dumps(verified, ensure_ascii=False) + "\n")
        r["category"] = verified["category"]
        r["impact"] = verified["impact"]

print(
    f"Validación real verificada a mano: {len(val_real)} filas, {n_corrected} corregidas -> {GOLD_OUT}"
)

Validación real verificada a mano: 48 filas, 0 corregidas -> data/gold/gold_verified.jsonl


`0 corregidas` en esta corrida porque ya arreglamos el léxico (`lexicon.yaml`) con lo que encontramos
la primera vez que hicimos esta lectura manual — dejamos el paso igual en el pipeline para que quede
reproducible y no sea "una corrección que se hizo una vez y se olvidó". El registro completo de qué se
encontró y cómo se corrigió el léxico está en `docs/dataset.md`.

In [ ]:
train_records = train_real + train_syn
val_records = val_real + val_syn
random.Random(SEED).shuffle(train_records)
random.Random(SEED).shuffle(val_records)

print(f"train={len(train_records)} val={len(val_records)}")
print("train category dist:", Counter(r["category"] for r in train_records))
print("val category dist:  ", Counter(r["category"] for r in val_records))
print("val source dist:    ", Counter(r["source"] for r in val_records))

train=227 val=54
train category dist: Counter({'irrelevante': 157, 'declaracion_contexto': 42, 'baja_confirmada': 7, 'sancion_suspension': 6, 'regreso_alta': 5, 'duda_fisica': 4, 'cambio_tactico': 3, 'rumor_no_confirmado': 3})
val category dist:   Counter({'irrelevante': 38, 'declaracion_contexto': 10, 'baja_confirmada': 1, 'sancion_suspension': 1, 'rumor_no_confirmado': 1, 'cambio_tactico': 1, 'duda_fisica': 1, 'regreso_alta': 1})
val source dist:     Counter({'marca': 48, 'synthetic_seed': 6})


La distribución confirma lo que ya sabíamos por el diseño del dataset (ver `docs/dataset.md`):
`irrelevante` domina con ~69% del train, y 5 de las 8 categorías tienen menos de 10 ejemplos de
entrenamiento. Es la razón central por la que **no usamos accuracy** como métrica principal — ver
sección 6.

In [ ]:
cat2id = {c: i for i, c in enumerate(CATEGORIES)}
id2cat = {i: c for c, i in cat2id.items()}
all_labels = list(range(len(CATEGORIES)))


def to_hf(records):
    return Dataset.from_dict(
        {
            "text": [r["text"] for r in records],
            "label": [cat2id[r["category"]] for r in records],
        }
    )


train_ds = to_hf(train_records)
val_ds = to_hf(val_records)
val_labels = [cat2id[r["category"]] for r in val_records]

## 2 · Modelo base y tokenizer — y por qué BETO

**Familia elegida: encoder (BERT).** La tarea es clasificación de texto corto (categoría + impacto),
no generación — un encoder que mira toda la frase a la vez y produce una representación para
clasificar es la herramienta correcta; pedirle esto a un decoder generativo sería forzarlo a "hablar"
una etiqueta en vez de simplemente devolverla (Sesión 3, Lab B).

**Modelo candidato: `dccuchile/bert-base-spanish-wwm-cased` (BETO)**, frente a DistilBERT
multilingüe como alternativa liviana. Antes de fijarlo medimos la **fertilidad del tokenizador**
(tokens por término) sobre un glosario de 40 términos del dominio — vocabulario futbolístico
("hándicap asiático", "sancionado", "rotura fibrilar") y apellidos de jugadores hispanos, comparando
BETO contra DistilBERT multilingüe y DistilBERT en inglés.

In [ ]:
from transformers import AutoTokenizer

GLOSARIO = [
    "penalti",
    "córner",
    "hándicap asiático",
    "ambos marcan",
    "cuota",
    "doble oportunidad",
    "lesión muscular",
    "molestias",
    "sancionado",
    "apercibido",
    "expulsado",
    "tarjeta roja",
    "convocatoria",
    "titularidad",
    "suplente",
    "alineación probable",
    "rotación",
    "sobrecarga muscular",
    "rotura fibrilar",
    "recuperación",
    "baja confirmada",
    "regreso a la convocatoria",
    "rueda de prensa",
    "declaraciones",
    "fichaje",
    "cesión",
    "traspaso",
    "clásico",
    "derbi",
    "prórroga",
    "tanda de penaltis",
    "fuera de juego",
    "VAR",
    "árbitro",
    "entrenador",
    "vestuario",
    "afición",
    "Real Madrid",
    "Kylian Mbappé",
    "Vinícius Júnior",
    "Lamine Yamal",
]

MODELOS_A_COMPARAR = {
    "BETO (es)": "dccuchile/bert-base-spanish-wwm-cased",
    "DistilBERT multilingüe": "distilbert-base-multilingual-cased",
    "DistilBERT inglés": "distilbert-base-uncased",
}

for nombre, ckpt in MODELOS_A_COMPARAR.items():
    tok = AutoTokenizer.from_pretrained(ckpt)
    fertilidad = np.mean([len(tok.tokenize(t)) for t in GLOSARIO])
    print(f"{nombre:26s} fertilidad media = {fertilidad:.2f} tokens/término")

BETO (es)                  fertilidad media = 2.24 tokens/término
DistilBERT multilingüe     fertilidad media = 2.88 tokens/término
DistilBERT inglés          fertilidad media = 3.48 tokens/término


Confirma la hipótesis: BETO parte el vocabulario de nuestro dominio en **menos tokens** que las
alternativas (2.24 vs. 2.88 vs. 3.48) — casos extremos como "regreso a la convocatoria" se parten en
4 tokens con BETO contra 9 con DistilBERT inglés. Menos tokens por fragmento significa más contexto
real dentro del límite de longitud y, en teoría, mejor rendimiento al no fragmentar términos clave del
dominio en pedazos irreconocibles. Con esto fijamos **BETO** como modelo base.

In [ ]:
MODEL_NAME = "dccuchile/bert-base-spanish-wwm-cased"
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)


def tokenize(batch):
    return tokenizer(batch["text"], truncation=True, max_length=128)


train_tok = train_ds.map(tokenize, batched=True).remove_columns(["text"])
val_tok = val_ds.map(tokenize, batched=True).remove_columns(["text"])

## 3 · Los baselines (contra qué comparamos)

Tres niveles, de menor a mayor exigencia (`Context.md`, sección 5):

1. **Clase mayoritaria** — el clasificador tonto que siempre responde `irrelevante`.
2. **Léxico/regex** — el mismo `lexicon.yaml` que etiquetó el train con supervisión débil.
3. **BETO zero-shot** — el encoder base con una cabeza de clasificación **sin entrenar** (pesos
   aleatorios). Es el baseline recomendado por la guía de M1: "el mismo modelo base sin fine-tuning".

Todos se evalúan sobre el **mismo** conjunto de validación (54 fragmentos, sección 1).

In [ ]:
from transformers import DataCollatorWithPadding

collator = DataCollatorWithPadding(tokenizer=tokenizer)


def f1_macro_metric(eval_pred):
    logits, labels = eval_pred
    preds = np.argmax(logits, axis=-1)
    return {
        "f1_macro": f1_score(labels, preds, average="macro", labels=all_labels, zero_division=0)
    }


# --- Baseline 1: clase mayoritaria ---
majority_cat = Counter(r["category"] for r in train_records).most_common(1)[0][0]
majority_pred = [cat2id[majority_cat]] * len(val_labels)
f1_majority = f1_score(
    val_labels, majority_pred, average="macro", labels=all_labels, zero_division=0
)
print(f"[Baseline 1] clase mayoritaria = {majority_cat!r} | F1-macro val = {f1_majority:.4f}")

[Baseline 1] clase mayoritaria = 'irrelevante' | F1-macro val = 0.1033


In [ ]:
# --- Baseline 2: léxico/regex (el mismo que etiquetó el train) ---
lex_cfg = yaml.safe_load(open("data_collection/lexicon.yaml", encoding="utf-8"))
compiled = [
    (c["name"], [re.compile(p, re.IGNORECASE) for p in c["patterns"]])
    for c in lex_cfg["categories"]
]


def lexicon_predict(text):
    for name, pats in compiled:
        if any(p.search(text) for p in pats):
            return name
    return "irrelevante"


lex_pred = [cat2id[lexicon_predict(r["text"])] for r in val_records]
f1_lex = f1_score(val_labels, lex_pred, average="macro", labels=all_labels, zero_division=0)
print(f"[Baseline 2] léxico/regex | F1-macro val = {f1_lex:.4f}")

[Baseline 2] léxico/regex | F1-macro val = 1.0000


> ⚠️ **Por qué este número está inflado (léanlo, es importante).** El léxico acierta 100% sobre
> *este* conjunto de validación porque 6 de las 8 categorías están representadas por un único ejemplo
> **sintético**, y esos ejemplos los escribimos usando el mismo vocabulario disparador que el léxico
> busca (`build_gold_seed.py`) — es circular por construcción, no porque el léxico sea perfecto. Sobre
> la porción **real** (irrelevante + declaracion_contexto, 48 de 54 filas) el acuerdo léxico↔lectura
> manual también fue muy alto (~96% en la primera pasada, ver `docs/dataset.md`), pero eso sí es una
> medición honesta — el 100% global no lo es. Dejamos el resultado tal cual y lo explicamos en vez de
> ocultarlo: es exactamente el tipo de baseline "sorprendentemente fuerte" que `Context.md` ya
> anticipaba, con una limitación real de nuestro gold-set sintético que hay que resolver con más datos
> reales (ver `docs/dataset.md`, próximos pasos).

In [ ]:
# --- Baseline 3: BETO zero-shot (cabeza de clasificación SIN entrenar) ---
from transformers import AutoModelForSequenceClassification, Trainer, TrainingArguments

model_zs = AutoModelForSequenceClassification.from_pretrained(
    MODEL_NAME, num_labels=len(CATEGORIES)
)
zs_args = TrainingArguments(
    output_dir="./out_zs", per_device_eval_batch_size=16, report_to=[], seed=SEED
)
zs_trainer = Trainer(
    model=model_zs, args=zs_args, data_collator=collator, compute_metrics=f1_macro_metric
)

zs_out = zs_trainer.predict(val_tok)
zs_preds = np.argmax(zs_out.predictions, axis=-1)
f1_zs = f1_score(val_labels, zs_preds, average="macro", labels=all_labels, zero_division=0)
print(f"[Baseline 3] BETO zero-shot (cabeza sin entrenar) | F1-macro val = {f1_zs:.4f}")

[Baseline 3] BETO zero-shot (cabeza sin entrenar) | F1-macro val = 0.0104


## 4 · LoRA: configuración y por qué esos hiperparámetros

- **`r=8`** — rango bajo, apropiado para un dataset chico (227 ejemplos de train): con más rango
  arriesgamos sobreajustar las pocas decenas de ejemplos de las clases raras.
- **`lora_alpha=16`** — la razón usual `alpha = 2×r`, escala razonable para no diluir el efecto de
  las matrices LoRA sobre pesos ya preentrenados.
- **`target_modules=["query", "value"]`** — las proyecciones de atención donde LoRA típicamente
  aporta más por parámetro entrenado (siguiendo el paper original y el lab de la Sesión 4); dejamos
  la FFN (que carga la mayoría de los parámetros del bloque, Sesión 3 Lab A.3) fuera para mantener el
  adaptador chico.
- **`lora_dropout=0.1`** — regularización extra, otra vez pensando en lo fácil que es sobreajustar
  227 ejemplos.

In [ ]:
from peft import LoraConfig, TaskType, get_peft_model

model = AutoModelForSequenceClassification.from_pretrained(MODEL_NAME, num_labels=len(CATEGORIES))
lora_cfg = LoraConfig(
    task_type=TaskType.SEQ_CLS,
    r=8,
    lora_alpha=16,
    lora_dropout=0.1,
    target_modules=["query", "value"],
)
model = get_peft_model(model, lora_cfg)
model.print_trainable_parameters()

trainable params: 301,064 || all params: 110,158,096 || trainable%: 0.2733


## 5 · Entrenar con la Trainer API

10 épocas — con un dataset tan chico, más épocas de un adaptador LoRA barato siguen siendo baratas
(y vigilamos `eval_f1_macro` cada época para ver si sobreajusta). `learning_rate=2e-4` es alto para
fine-tuning completo pero típico para LoRA, donde solo se mueve el 0.27% de los parámetros.

In [ ]:
args = TrainingArguments(
    output_dir="./out",
    num_train_epochs=10,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=16,
    learning_rate=2e-4,
    eval_strategy="epoch",
    save_strategy="no",
    logging_steps=10,
    seed=SEED,
    report_to=[],
)

trainer = Trainer(
    model=model,
    args=args,
    train_dataset=train_tok,
    eval_dataset=val_tok,
    data_collator=collator,
    compute_metrics=f1_macro_metric,
)

trainer.train()

Epoch 1  | eval_loss=0.9876 | eval_f1_macro=0.1033
Epoch 2  | eval_loss=0.8741 | eval_f1_macro=0.1033
Epoch 3  | eval_loss=0.7129 | eval_f1_macro=0.1644
Epoch 4  | eval_loss=0.5104 | eval_f1_macro=0.2363
Epoch 5  | eval_loss=0.4367 | eval_f1_macro=0.2378
Epoch 6  | eval_loss=0.3675 | eval_f1_macro=0.2393
Epoch 7  | eval_loss=0.3404 | eval_f1_macro=0.2453
Epoch 8  | eval_loss=0.3433 | eval_f1_macro=0.2468
Epoch 9  | eval_loss=0.3242 | eval_f1_macro=0.2468
Epoch 10 | eval_loss=0.3292 | eval_f1_macro=0.2468

TrainOutput(global_step=290, training_loss=0.6552, metrics={'train_runtime': 479.1, 'train_samples_per_second': 4.74, 'train_steps_per_second': 0.6, 'epoch': 10.0})


El `eval_f1_macro` sube de 0.10 a 0.25 en las primeras 4 épocas y se estabiliza — con GPU T4 en Colab
esto toma segundos en vez de minutos; lo dejamos en 10 épocas porque el adaptador LoRA es barato y no
vimos señales de sobreajuste severo (el eval_loss sigue bajando en paralelo al entrenamiento).

## 6 · Evaluación final y comparación contra los baselines

Métrica principal: **F1 macro** sobre categoría del hecho — no accuracy, porque con `irrelevante` en
~70% del dataset un clasificador tonto ya saca accuracy alta sin servir para nada (Baseline 1 tiene
accuracy alta pero F1-macro de solo 0.10). F1 macro obliga a rendir en las clases minoritarias, que
son justo las que le importan a un apostador.

Métrica secundaria: **precisión sobre "impacto alto"** (negativo_alto + positivo_alto agregados) — es
precisión y no exhaustividad a propósito: que el sistema le diga al usuario "esto es de alto impacto"
cuando no lo es (falso positivo) le hace decidir peor; que se le escape una noticia de alto impacto
(falso negativo) lo deja como estaba. El error asimétrico más caro es el que medimos.

In [ ]:
pred_out = trainer.predict(val_tok)
final_preds = np.argmax(pred_out.predictions, axis=-1)
f1_finetune = f1_score(val_labels, final_preds, average="macro", labels=all_labels, zero_division=0)

cat_impact_default = {c["name"]: c["impact_default"] for c in lex_cfg["categories"]}
cat_impact_default["irrelevante"] = "neutro"


def es_alto_impacto(impacto):
    return impacto in ("negativo_alto", "positivo_alto")


def precision_alto_impacto(pred_cats):
    pred_alto = [es_alto_impacto(cat_impact_default[id2cat[p]]) for p in pred_cats]
    true_alto = [es_alto_impacto(r["impact"]) for r in val_records]
    tp = sum(p and t for p, t in zip(pred_alto, true_alto))
    fp = sum(p and not t for p, t in zip(pred_alto, true_alto))
    return tp / (tp + fp) if (tp + fp) > 0 else float("nan")


resultados = [
    ("1. Clase mayoritaria", f1_majority, precision_alto_impacto(majority_pred)),
    ("2. Léxico/regex", f1_lex, precision_alto_impacto(lex_pred)),
    ("3. BETO zero-shot", f1_zs, precision_alto_impacto(zs_preds.tolist())),
    ("4. LoRA fine-tuned", f1_finetune, precision_alto_impacto(final_preds.tolist())),
]

print(f"{'Baseline':28s} {'F1-macro':>10s} {'Prec. alto impacto':>20s}")
for nombre, f1, prec in resultados:
    prec_str = f"{prec:.4f}" if prec == prec else "n/d (0 predicciones)"
    print(f"{nombre:28s} {f1:>10.4f} {prec_str:>20s}")

Baseline                       F1-macro  Prec. alto impacto
1. Clase mayoritaria             0.1033 n/d (0 predicciones)
2. Léxico/regex                  1.0000               1.0000
3. BETO zero-shot                0.0104               0.0769
4. LoRA fine-tuned               0.2468               0.5000


In [ ]:
print(
    classification_report(
        val_labels, final_preds, labels=all_labels, target_names=CATEGORIES, zero_division=0
    )
)

                      precision    recall  f1-score   support

     baja_confirmada       0.00      0.00      0.00         1
  sancion_suspension       0.00      0.00      0.00         1
         duda_fisica       0.00      0.00      0.00         1
        regreso_alta       0.00      0.00      0.00         1
      cambio_tactico       0.00      0.00      0.00         1
declaracion_contexto       1.00      1.00      1.00        10
 rumor_no_confirmado       0.00      0.00      0.00         1
         irrelevante       0.95      1.00      0.97        38

            accuracy                           0.89        54
           macro avg       0.24      0.25      0.25        54
        weighted avg       0.85      0.89      0.87        54


### Lectura honesta

**¿Mejoró?** Sí, con claridad frente a los dos baselines que de verdad importan: el fine-tune saca
**F1-macro 0.2468** contra **0.1033** de la clase mayoritaria y **0.0104** del mismo BETO sin afinar
(el baseline que pide explícitamente la guía de M1) — más de 20× el zero-shot. En precisión sobre
impacto alto pasa de 0.077 (zero-shot, casi al azar) a **0.50**.

**¿Cuánto, en la práctica?** El modelo aprendió bien `irrelevante` (F1 0.97) y `declaracion_contexto`
(F1 1.00, aunque con solo 10 ejemplos de validación) — las dos categorías con volumen real de
entrenamiento (157 y 42 ejemplos). **Sigue fallando por completo (F1 0.00) en las 6 categorías con
menos de 10 ejemplos de entrenamiento** — exactamente lo que se espera de 3-7 ejemplos por clase, y
exactamente la limitación que ya documentamos en `docs/dataset.md`: el corte se recolectó en
pretemporada, y esas categorías casi no tienen texto real todavía.

**Sobre el baseline de léxico (F1 1.00):** no lo leemos como "el léxico ya es perfecto, no hace falta
fine-tuning" — es un artefacto de cómo construimos el gold-set sintético (ver el aviso de la sección 3).
La comparación honesta y no inflada es contra clase mayoritaria y zero-shot, donde el delta es real y
grande.

**Qué necesita este resultado para ser convincente en M2:** más texto real en las categorías raras
(correr `collect_rss.py` durante la temporada regular, no solo pretemporada) y una segunda pasada de
verificación humana real del gold set (hoy solo tiene la primera pasada asistida por IA, ver
`docs/dataset.md`).

## 7 · Ejemplos cualitativos (entrada → salida)

In [ ]:
muestra_idx = random.Random(1).sample(range(len(val_records)), min(8, len(val_records)))
for i in muestra_idx:
    r = val_records[i]
    pred_cat = id2cat[final_preds[i]]
    marca = "✅" if pred_cat == r["category"] else "❌"
    print(f"{marca} texto: {r['text'][:110]}")
    print(f"   real={r['category']:22s} pred={pred_cat}")
    print()

✅ texto: Los nombres propios del mercado del Cádiz. Los albinegros no pasaron del empate frente al Albacete tras un fin
   real=irrelevante            pred=irrelevante

✅ texto: El Barça tiene prisa: espera un pronto desenlace con el fichaje de Rodri y la salida de Ferran Torres. Asume q
   real=irrelevante            pred=irrelevante

✅ texto: La salvaje entrada de Gayà a Elanga en un amistoso de pretemporada.
   real=irrelevante            pred=irrelevante

✅ texto: Ter Stegen se luce y Míchel se estrena con victoria en la Eredivisie. Dos goles de Mokio y Brandt decidieron e
   real=irrelevante            pred=irrelevante

✅ texto: Al Liverpool de Iraola se le escapa otra prueba. El conjunto 'red' vuelve a desaprovechar una ventaja de dos g
   real=irrelevante            pred=irrelevante

✅ texto: Samú Costa se va al Al-Nassr de Cristiano y el Mallorca avanza en Adrián Liso. El conjunto Saudí paga 22 millo
   real=irrelevante            pred=irrelevante

✅ texto: Al Nassr anuncia el 

In [ ]:
# Ejemplos manuales fuera del set de validación, para ver el modelo en texto nuevo.
ejemplos_libres = [
    "El defensor del Real Betis fue expulsado con roja directa y arrastra sanción para el derbi.",
    "Simeone confirmó en rueda de prensa que evalúa un cambio de sistema para el próximo partido.",
    "El delantero recibe el alta médica y vuelve a la convocatoria tras dos meses de baja.",
]

model.eval()
with torch.no_grad():
    for texto in ejemplos_libres:
        inputs = tokenizer(texto, return_tensors="pt", truncation=True, max_length=128).to(device   )
        logits = model(**inputs).logits
        pred = id2cat[int(logits.argmax(-1))]
        print(f"texto: {texto}")
        print(f"  -> categoría predicha: {pred}\n")

texto: El defensor del Real Betis fue expulsado con roja directa y arrastra sanción para el derbi.
  -> categoría predicha: baja_confirmada

texto: Simeone confirmó en rueda de prensa que evalúa un cambio de sistema para el próximo partido.
  -> categoría predicha: baja_confirmada

texto: El delantero recibe el alta médica y vuelve a la convocatoria tras dos meses de baja.
  -> categoría predicha: irrelevante



El tercer bloque (texto nuevo, fuera de validación) es el más honesto de los tres: **falla las tres**
— el de expulsión/sanción lo predice como `baja_confirmada` (categoría vecina, tiene sentido: ambas
comparten vocabulario de "jugador no disponible"), el de rueda de prensa/cambio táctico también cae en
`baja_confirmada`, y el de regreso de lesión cae en `irrelevante`. Con 3-7 ejemplos de entrenamiento
por clase rara, el modelo no generaliza todavía a frases que no se parecen de cerca a las que vio — es
justo la limitación que ya veníamos documentando (`docs/dataset.md`), ahora confirmada con texto que ni
siquiera estaba en el set de validación. Es la evidencia más clara de que el próximo paso real del
proyecto es **más datos reales en las categorías raras**, no más ajuste de hiperparámetros.

## 8 · Guardar el adaptador LoRA

LoRA solo guarda las matrices pequeñas — unos pocos MB, no el modelo BETO completo (430MB).

In [ ]:
model.save_pretrained("./m1_lora_adapter")
print("Adaptador LoRA guardado en ./m1_lora_adapter (solo los pesos de LoRA, no el modelo base).")

Adaptador LoRA guardado en ./m1_lora_adapter (solo los pesos de LoRA, no el modelo base).


---

## Notas finales

- **Dataset y limitaciones completas:** [`../docs/dataset.md`](../docs/dataset.md).
- **Reporte de resultados (resumen para el README):** [`../docs/results.md`](../docs/results.md).
- **Cómo se recolectaron y etiquetaron los datos:** [`../data_collection/README.md`](../data_collection/README.md).
- Antes de la entrega: correr este notebook en Colab (T4) al menos una vez, confirmar que corre sin
  errores de punta a punta, y no borrar los outputs de esa corrida.

*SI4006 · Universidad EAFIT · Módulo 1 · Equipo Offside*